File: SWEA/[7465] 창용 마을 무리의 개수/main.java
import java.util.*;
import java.io.*;

class Solution
{
    static ArrayList<ArrayList<Integer>> linked;
    static boolean[] visited;

	public static void main(String args[]) throws Exception
	{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
		int T = Integer.parseInt(st.nextToken());
		
		for(int test_case = 1; test_case <= T; test_case++)
		{
            st = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(st.nextToken());
            int m = Integer.parseInt(st.nextToken());
            linked = new ArrayList<>();
            // n+1 -> 1부터 시작하기
            visited = new boolean[n + 1];
            
            // 초기화...
            for (int i = 0; i <= n; i++) {
                linked.add(new ArrayList<>());
            }

            for (int i = 0; i < m; i++) {
                st = new StringTokenizer(br.readLine());
                int a = Integer.parseInt(st.nextToken());
                int b = Integer.parseInt(st.nextToken());
                // 아는 사람들 입력 -> linkedList
                linked.get(a).add(b);
                linked.get(b).add(a);
            }       

            // 건너 건너 아는 사람이면 같은 무리 -> DFS

            int cnt = 0;
            for (int i = 1; i <= n; i++){
                if(!visited[i]){
                    visited[i] = true;
                    cnt++;
                    dfs(i);
                }
            }

            System.out.println("#" + test_case + " " + cnt);
		}
	}
    static void dfs(int cur){
        for(int neighborIDX : linked.get(cur)){
            if(!visited[neighborIDX]){
                visited[neighborIDX] = true;
                dfs(neighborIDX);
            }
        }
    }

}

File: SWEA/[7465] 창용 마을 무리의 개수/main.py
T = int(input())

# import sys


# input = sys.stdin.read
# data = input().strip().split("\n")
def dfs(cur):
    for neighborIDX in linked[cur]:
        if not visited[neighborIDX]:
            visited[neighborIDX] = True
            dfs(neighborIDX)


# 여러개의 테스트 케이스가 주어지므로, 각각을 처리합니다.
for test_case in range(1, T + 1):
    n, m = map(int, input().split())

    # n+1 -> 1부터 시작하기
    linked = [[] for _ in range(n + 1)]
    visited = [False] * (n + 1)
    for i in range(m):
        # 아는 사람들 입력 -> linkedList
        a, b = map(int, input().split())
        linked[a].append(b)
        linked[b].append(a)

    # 건너 건너 아는 사람이면 같은 무리 -> DFS
    cnt = 0
    for i in range(1, n + 1):
        if not visited[i]:
            visited[i] = True
            cnt += 1
            dfs(i)

    print(f"#{test_case} {cnt}")
